File: Code/DynamicProgramming/LIS/Largest_Divisible_Subset_368.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution
{
public:
    vector<int> largestDivisibleSubset(vector<int> nums)
    {
        // longest incresing subsequence
        sort(nums.begin(), nums.end());
        vector<int> dp[nums.size()];
        int idx = 0;
        for (int i = 0; i < nums.size(); i++)
        {
            for (int j = 0; j <= i; j++)
            {
                if (nums[i] % nums[j] == 0)
                {
                    if (dp[i].size() <= dp[j].size())
                    {
                        dp[i] = dp[j];
                    }
                }
            }
            dp[i].push_back(nums[i]);
            if (dp[i].size() >= dp[idx].size())
                idx = i;
        }
        return dp[idx];
    }
};
int main()
{
#ifndef ONLINE_JUDGE
    freopen("input.txt", "r", stdin);
    freopen("output.txt", "w", stdout);
#endif
    Solution obj;
    auto ans = obj.largestDivisibleSubset({1, 2, 3});
    // cout << ans;
    return 0;
}

File: java/DynamicProgramming_java/LIS/Largest_Divisible_Subset_368.java
package java.DynamicProgramming_java.LIS;

import java.util.*;
import java.io.*;

// import static java.lang.Math.max;
// import static java.lang.Math.min;
// import static java.lang.Math.abs;
// import static java.lang.System.out;

public class Largest_Divisible_Subset_368 {

    public List<Integer> largestDivisibleSubset(int[] nums) {
        // longest incresing subsequence
        Arrays.sort(nums);
        int dp[] = new int[nums.length];
        int idx = 0;
        for (int i = nums.length - 1; i >= 0; --i) {
            for (int j = nums.length - 1; j > i; --j) {
                if (nums[j] % nums[i] == 0) {
                    dp[i] = Math.max(dp[i], dp[j] + 1);
                }
            }
            if (dp[i] >= dp[idx])
                idx = i;
        }

        List<Integer> ans = new ArrayList<>();
        int prev = -1, maxInTotal = dp[idx];
        for (int i = idx; i < nums.length; i++) {
            if (dp[i] == maxInTotal && (nums[i] % prev == 0 || prev == -1)) {
                ans.add(nums[i]);
                prev = nums[i];
                maxInTotal--;
            }
        }

        return ans;
    }

    public static void main(String[] args) throws FileNotFoundException {
        File f = new File("output.txt");

        // Scanner ip = new Scanner(System.in);
        try (Scanner ip = new Scanner(new File("input.txt"));
                PrintWriter op = new PrintWriter(f);) {
            Largest_Divisible_Subset_368 obj = new Largest_Divisible_Subset_368();
            int grid[] = { 1, 2, 4, 8 };
            op.println(obj.largestDivisibleSubset(grid));

        }
    }

}
